istioctl/pkg/kubernetes/client.go

```diff
@@ -217,3 +223,84 @@ func (client *Client) GetPilotAgentContainer(podName, podNamespace string) (stri
 	}
 	return proxyContainer, nil
 }
+
+type podDetail struct {
+	binary    string
+	container string
+}
+
+// GetIstioVersions gets the version for each Istio component
+func (client *Client) GetIstioVersions(namespace string) (*version.MeshInfo, error) {
+	pods, err := client.GetIstioPods(namespace, map[string]string{"labelSelector": "istio"})
+	if err != nil {
+		return nil, err
+	}
+	if len(pods) == 0 {
+		return nil, errors.New("unable to find any Istio pod in namespace " + namespace)
+	}
+
+	labelToPodDetail := map[string]podDetail{
+		"pilot":            {"/usr/local/bin/pilot-discovery", "discovery"},
+		"citadel":          {"/usr/local/bin/istio_ca", "citadel"},
+		"egressgateway":    {"/usr/local/bin/pilot-agent", "egressgateway"},
+		"galley":           {"/usr/local/bin/galley", "validator"},
+		"ingressgateway":   {"/usr/local/bin/pilot-agent", "ingressgateway"},
+		"telemetry":        {"/usr/local/bin/mixs", "mixer"},
+		"policy":           {"/usr/local/bin/mixs", "mixer"},
+		"sidecar-injector": {"/usr/local/bin/sidecar-injector", "sidecar-injector-webhook"},
+	}
+
+	res := version.MeshInfo{}
+	for _, pod := range pods {
+		component := pod.Labels["istio"]
+
+		// Special cases
+		switch component {
+		case "statsd-prom-bridge":
+			continue
+		case "mixer":
+			component = pod.Labels["istio-mixer-type"]
+		}
+
+		server := version.ServerInfo{Component: component}
+
+		if detail, ok := labelToPodDetail[component]; ok {
+			cmd := []string{detail.binary, "version"}
+			cmdJSON := append(cmd, "-o", "json")
+
+			var stdout, stderr *bytes.Buffer
+			var err error
+			var info version.BuildInfo
+
+			stdout, stderr, err = client.PodExec(pod.Name, pod.Namespace, detail.container, cmdJSON)
```

Comment:
@liamawhite done. Although I replaced just one call. The other one I couldn't because I'm relying on the `stderr` in order to know whether we are dealing with old server components - which don't recognize the `--output` flag. That way we are backward compatible, meaning newer `istioctl` can get version info from old server components.